File: ipc-codegen/examples/ts/echo/echo_server.ts
/**
 * Echo IPC server (TypeScript) — uses GENERATED dispatch + IPC server template.
 * Usage: npx tsx echo_server.ts --socket /tmp/echo.sock
 */
import { createServer } from './generated/ipc_server.js';
import { dispatch } from './generated/server.js';
import type { Handler } from './generated/server.js';
import type {
  EchoBytes, EchoBytesResponse,
  EchoFields, EchoFieldsResponse,
  EchoNested, EchoNestedResponse,
} from './generated/echo_types.js';

const args = process.argv.slice(2);
const socketIdx = args.indexOf('--socket');
const socketPath = socketIdx >= 0 ? args[socketIdx + 1] : undefined;
if (!socketPath) {
  console.error('Usage: echo_server.ts --socket <path>');
  process.exit(1);
}

// Implement the GENERATED Handler interface — echo everything back
const handler: Handler = {
  async echoBytes(cmd: EchoBytes): Promise<EchoBytesResponse> {
    return { data: cmd.data };
  },
  async echoFields(cmd: EchoFields): Promise<EchoFieldsResponse> {
    return { a: cmd.a, b: cmd.b, name: cmd.name };
  },
  async echoNested(cmd: EchoNested): Promise<EchoNestedResponse> {
    return { inner: cmd.inner };
  },
};

createServer(socketPath, (commandName, payload) => dispatch(handler, commandName, payload));

File: ipc-codegen/examples/rust/echo/src/echo_server.rs
//! Echo IPC server — uses GENERATED dispatch + server template + types.
//! Usage: echo_server --socket /tmp/echo.sock

use echo_wire_compat::generated::echo_server::Handler;
use echo_wire_compat::generated::echo_types::*;
use echo_wire_compat::generated::error::{IpcError, Result};
use echo_wire_compat::generated::ipc_server;
use std::cell::RefCell;

struct EchoHandler;

impl Handler for EchoHandler {
    fn bytes(&mut self, cmd: EchoBytes) -> Result<EchoBytesResponse> {
        Ok(EchoBytesResponse { data: cmd.data })
    }
    fn fields(&mut self, cmd: EchoFields) -> Result<EchoFieldsResponse> {
        Ok(EchoFieldsResponse { a: cmd.a, b: cmd.b, name: cmd.name })
    }
    fn nested(&mut self, cmd: EchoNested) -> Result<EchoNestedResponse> {
        Ok(EchoNestedResponse { inner: cmd.inner })
    }
}

fn main() -> Result<()> {
    let args: Vec<String> = std::env::args().collect();
    let socket_path = args.iter()
        .position(|a| a == "--socket")
        .and_then(|i| args.get(i + 1))
        .expect("Usage: echo_server --socket <path>");

    let _ = std::fs::remove_file(socket_path);

    // Wrap handler in RefCell so the Fn closure can borrow mutably.
    let handler = RefCell::new(EchoHandler);

    ipc_server::serve(socket_path, |payload: &[u8]| {
        // Deserialize: [Command]
        let request: Vec<Command> = rmp_serde::from_slice(payload)
            .unwrap_or_default();

        let command = match request.into_iter().next() {
            Some(cmd) => cmd,
            None => {
                let err = Response::EchoErrorResponse(EchoErrorResponse {
                    message: "empty request".to_string(),
                });
                return rmp_serde::to_vec_named(&err).unwrap_or_default();
            }
        };

        // Check for shutdown before dispatch
        let is_shutdown = matches!(&command, Command::EchoShutdown(_));

        let response = match echo_wire_compat::generated::echo_server::dispatch(
            &mut *handler.borrow_mut(), command
        ) {
            Ok(resp) => resp,
            Err(_e) => {
                if is_shutdown {
                    Response::EchoShutdownResponse(EchoShutdownResponse {})
                } else {
                    Response::EchoErrorResponse(EchoErrorResponse {
                        message: _e.to_string(),
                    })
                }
            }
        };

        rmp_serde::to_vec_named(&response).unwrap_or_default()
    }).map_err(|e| IpcError::Io(e))
}
